### PR TITLE
FSRV 3314

### PR DESCRIPTION
**Pull request by wilig**
Add authentication/authorization to the iiif service

### CODE REVIEW

File: loris/ithaka.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    out = base64.urlsafe_b64decode(session_text+"===")
+    data = json.loads(gzip.GzipFile(fileobj=StringIO(out)).read())
+    return data['id']
+
+
+class DiscoverableService:
+
+    @staticmethod
+    def resolve(service_name):
+        eureka_url = "https://eureka.apps.%s.cirrostratus.org" % os.getenv("SGK_ENVIRONMENT", "test")
```
Comment (daneah) [inline loris/ithaka.py]: I think you can avoid a hit to apps-gateway by foregoing part of the domain here:

```suggestion
        eureka_url = "https://eureka.%s.cirrostratus.org" % os.getenv("SGK_ENVIRONMENT", "test")
```

Comment (daneah) [reply loris/ithaka.py]: I take it this app doesn't have a Eureka sidecar of its own?

File: tests/authorizer_t.py [hunk 2]
```diff
@@ -343,3 +360,170 @@ def test_salt_must_be_bytes(self):
...
+
+class TestIthakaAuthorizer():
+
+    @pytest.fixture
+    def authorizer(self):
+        return IthakaAuthorizer({})
+
+    @pytest.fixture
+    def okay_info(self):
+        return ImageInfo(None, "67352ccc-d1b0-11e1-89ae-279075081939.jp2",\
```
Comment (daneah) [inline tests/authorizer_t.py]: General comment about readability (it's consistent in most of the forum code, so I wouldn't necessarily say to follow this suggestion _over_ consistency, but maybe for new stuff): if arguments don't fit on one line, each on a separate line is more readable:

```python
    @pytest.fixture
    def okay_info(self):
        return ImageInfo(
            None,
            "67352ccc-d1b0-11e1-89ae-279075081939.jp2",
            "img/67352ccc-d1b0-11e1-89ae-279075081939.jp2",
            "jp2",
        )
```

Shows up better in a diff when a single arg changes, makes it less awkward to leave a hanging comma for new arguments are added, and avoids having to parse each line to understand if there is one or more than one argument present!

It's generally more readable like the above than the below, as well, especially for lines with long method names that would push all the args off to the right.

```python
    @pytest.fixture
    def okay_info(self):
        return ImageInfo(None,
                         "67352ccc-d1b0-11e1-89ae-279075081939.jp2",
                         "img/67352ccc-d1b0-11e1-89ae-279075081939.jp2",
                         "jp2")
```

Comment (daneah) [reply tests/authorizer_t.py]: Case in point: I missed that the last `"jp2"` was its own argument while writing this suggestion originally 😄

File: tests/ithaka_t.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    result = dict(status=True)
+    replace_request(MockResponse(json=result))
+    assert (result == IACService.session('12345', ['123', '456']))
+
+
+def test_eme_service(replace_resolve, replace_request):
+    replace_resolve(EMEService)
+    result = dict(status=True)
+    replace_request(MockResponse(json=result))
+    assert (result == EMEService.get_tokens('doi/5'))
```
Comment (daneah) [inline tests/ithaka_t.py]: The parens around these asserted expressions are redundant, yeah?

File: loris/authorizer.py [hunk 4]
```diff
@@ -407,3 +411,85 @@ def get_services_info(self, info):
             'fp': info.src_img_fp,
         }
         requests.post(self.services_url, data=data)
+
+
+class IthakaAuthorizer(_AbstractAuthorizer):
```
Comment (daneah) [inline loris/authorizer.py]: This looks about right to me, and I wish we had something so simple over in ithaka/jstor 🙂

File: loris/ithaka.py (hunk 1 above)
Comment (wilig) [reply loris/ithaka.py]: Doesn't seem to.

File: tests/authorizer_t.py (hunk 2 above)
Comment (wilig) [reply tests/authorizer_t.py]: Noted.  I'll update them.  Would the black formatting utility enforce this as well?

File: tests/ithaka_t.py (hunk 3 above)
Comment (wilig) [reply tests/ithaka_t.py]: Lispiness slipping in

File: loris/ithaka.py (hunk 1 above)
Comment (daneah) [reply loris/ithaka.py]: If you expect service discovery to happen frequently, you may want to talk with CORE about the intricacies of having a sidecar vs. calling Eureka directly...I don't rightly know much in that arena, though my guess is that making `localhost` calls would be beneficial!

File: tests/authorizer_t.py (hunk 2 above)
Comment (daneah) [reply tests/authorizer_t.py]: Yep! Black would come out most often with the former. It occasionally does this, too:

```python
def some_function(
    arg_one, arg_two, another_one, good_golly, a_lot_of_these_innit
):
    ...
```

**Pull request merged**